### PR TITLE
refactor addon update nominated status

### LINKED ISSUE #19773: refactor `Addon.update_nominated_status` into `Addon.update_status`
follow-up for  #19728 - `update_nominated_status` was added to change the add-on status to `STATUS_NOMINATED` if there are any listed versions that are `STATUS_AWAITING_REVIEW` and the addon status is `STATUS_NULL`, but we should be able to **carefully** fold this functionality into `update_status` directly...  Also see thread https://github.com/mozilla/addons-server/pull/19728#discussion_r984674335

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #19773 - but restating what I noted in the issue:
- we've got to stop changing the add-on status manually for this to be a complete fix, really.  Which is a massive task when a) our tests do it **everywhere** and b) we still need to set the status manually for `STATUS_DELETED` and `STATUS_DISABLED`.  That said, with the signals what pretty much happens is: if the status is manually set to the wrong state it gets reset to the correct state which just means extra database writes and unnecessary logging in our tests.
- the case of setting `STATUS_NOMINATED` and requiring complete metadata was unclear.  I went with the closest scenario to now, where an existing `STATUS_NULL` requires complete metadata to upgrade, but other statuses will downgrade to `STATUS_NOMINATED` without requiring complete metadata.

### CODE REVIEW

File: src/olympia/addons/tests/test_models.py [hunk 1]
```diff
@@ -2104,28 +2100,52 @@ def test_unlisted_versions_ignored(self):
...
+        addon = addon_factory(
+            status=amo.STATUS_NULL,
+            summary=None,
+            file_kw={'status': amo.STATUS_DISABLED},
+        )
+        assert addon.status == amo.STATUS_NULL
+
+        version_factory(addon=addon, file_kw={'status': amo.STATUS_AWAITING_REVIEW})
+        addon.reload()
+        # Because it don't have complete metadata the status isn't changed
```
Comment (bobsilverberg) [inline src/olympia/addons/tests/test_models.py:2123]: Nit: s/don't/doesn't

File: src/olympia/amo/tests/__init__.py [hunk 2]
```diff
@@ -644,6 +644,7 @@ def _get_created(created):
 
 def addon_factory(status=amo.STATUS_APPROVED, version_kw=None, file_kw=None, **kw):
     version_kw = version_kw or {}
+    file_kw = file_kw or {}
```
Comment (bobsilverberg) [inline src/olympia/amo/tests/__init__.py:647]: How is this different from specifying `file_kw={}` in the function definition?

File: src/olympia/devhub/utils.py [hunk 3]
```diff
@@ -350,6 +350,5 @@ def create_version_for_upload(addon, upload, channel, parsed_data=None):
         # The add-on's status will be STATUS_NULL when its first version is
         # created because the version has no files when it gets added and it
         # gets flagged as invalid. We need to manually set the status.
```
Comment (bobsilverberg) [inline src/olympia/devhub/utils.py:352]: Should this comment be changed in some way to indicate that we're not actually manually setting the status, but rather requesting that it be re-evaluated?

**Review by bobsilverberg — APPROVED** [review]
This looks good, but to be honest I am a bit dicey on all of the status setting rules/logic, so perhaps @diox should take a look at it as well? I also added some minor comments, so r+wc from me.

File: src/olympia/amo/tests/__init__.py (hunk 2 above)
Comment (eviljeff) [reply src/olympia/amo/tests/__init__.py:647]: it's a python gotcha - you can't use mutable defaults in the kwargs as they're shared with future calls to the function.

**Review by diox — APPROVED** [review]
> the case of setting `STATUS_NOMINATED` and requiring complete metadata was unclear. I went with the closest scenario to now, where an existing `STATUS_NULL` requires complete metadata to upgrade, but other statuses will downgrade to `STATUS_NOMINATED` without requiring complete metadata.

I'm not 100% sure what should happen in this case, but I think your approach is the safest. There might be a edge-case I'm not seeing, we'll see if that ever comes up...

**Pull request merged**